x/ibc/06-solomachine/update.go

```diff
@@ -0,0 +1,76 @@
+package solomachine
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	clientexported "github.com/cosmos/cosmos-sdk/x/ibc/02-client/exported"
+	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+)
+
+// CheckValidityAndUpdateState checks if the provided header is valid and updates
+// the consensus state if appropriate. It returns an error if:
+// - the client or header provided are not parseable to solo machine types
+// - the currently registered public key did not provide the update signature
+func CheckValidityAndUpdateState(
+	clientState clientexported.ClientState, header clientexported.Header,
+) (clientexported.ClientState, clientexported.ConsensusState, error) {
+	// cast the client state to solo machine
+	smClientState, ok := clientState.(ClientState)
+	if !ok {
+		return nil, nil, sdkerrors.Wrapf(
+			clienttypes.ErrInvalidClientType, "client state type %T is not solomachine", clientState,
+		)
+	}
+
+	smHeader, ok := header.(Header)
+	if !ok {
+		return nil, nil, sdkerrors.Wrapf(
+			clienttypes.ErrInvalidHeader, "header type %T is not solomachine", header,
+		)
+	}
+
+	if err := checkValidity(smClientState, smHeader); err != nil {
+		return nil, nil, err
+	}
+
+	smClientState, consensusState := update(smClientState, smHeader)
+	return smClientState, consensusState, nil
+}
+
+// checkValidity checks if the Solo Machine update signature is valid.
+func checkValidity(clientState ClientState, header Header) error {
+	// assert update sequence is current sequence
+	if header.Sequence != clientState.ConsensusState.Sequence {
```

Comment:
cc: @cwgoes the spec is wrong here: https://github.com/cosmos/ics/blob/master/spec/ics-006-solo-machine-client/README.md#validity-predicate

should say :

```typescript
assert(header.sequence === clientState.consensusState.sequence)
 
```

Reply:
Thanks - fixed in https://github.com/cosmos/ics/commit/494fb5a17cfb9bbc822cafa8852cdbcfd94793bf.